### DISCUSSION TITLE
threading

### DISCUSSION THREAD

**Discussion by RDP1974** [uNetworking/uWebSockets.js#761] [Q&A]
hi,
beginner with this lib,
I have big expertize in low level coding,
can I ask:
- I see how to dispatch many threads over the bind IP, also I see that many threads can be bind to the same unique IP, or that the app can be bind to many IP
- now those threads are firing the events as OnMessage(), my question is if I build a threadpool that after receive will do parsing of protocol commands, db queries, then these threads can call app send() or send broadcast without incur in race/synchro troubles?
- example:
   - main thread receive()
   - a worker thread parse the text, do a query, then call app.send() outside the main thread
 or we need protection as mutex or crt?

**Comment by RDP1974** [comment]
is perhaps safe and enough to use in the main flow async() with await() keywords?

**Comment by e3dio** [comment]
Each thread has own app and can call `app.publish` to its own connected sockets, if you want to send or publish to sockets connected to different threads you need to send messages between threads to tell it to broadcast to their sockets, Node has methods for sending messages between threads

**Reply by RDP1974** [reply]
Please can you provide an example? App should have a global dictionary of lookups ws, one unique publish list

**Reply by e3dio** [reply]
You are asking how to send messages between threads, read the docs for worker threads and BroadcastChannel https://nodejs.org/api/worker_threads.html

**Reply by RDP1974** [reply]
"Each thread has own app and can call app.publish to its own connected sockets"
so considering that's async io, will be easier to use 1 only main thread with a unique list of connected sockets,
then for parsing, computing, db queries, use async await functions (and a db pool to persist db connections).
Perhaps do you know when the main thread is bind to many IP, then the connected sockets is still unique?
Maybe for a big federated project will be necessary to use associative arrays to achieve user->server->thread->ws
any hint? thank you for your help

**Comment by RDP1974** [comment]
please help me in this

consider I dispatch many threads over app()

I use a map where I insert a object: thread app, ws

1) is correct to send a message one-to-many to use a foreach - app,ws,send()
2) can different workers call app,ws,send()? or the send should be done in the main thread?
3) ecma js 6 permits insert, delete, retrieve into maps from different workers?

thanks

**Reply by e3dio** [reply]
1. app.publish
2. send from thread that socket is connected to
3. map operations are done from thread that has the map

Send messages between threads as needed

**Reply by RDP1974** [reply]
I understand now, thank you

**Comment by RDP1974** [comment]
a question,
if I bind the APP to 2 IP, then those are threated into the main thread, or there we will have 2 threads, each with own set of sockets?

**Reply by ghost** [reply]
Read the readmore on the front page, this whole thread is a FAQ

**Comment by RDP1974** [comment]
App.listen(port, [](auto *listenSocket) {
    /* listenSocket is either nullptr or us_listen_socket */
})

ok, but, if I listen eg. port 3000 and 3001, then we will have 2 threads instead of 1?
the socketlist is unique? or each thread has its own list?

**Reply by ghost** [reply]
Why don't you just test it? Do a bit of your own research and easily find out

**Reply by RDP1974** [reply]
hi Alex,
ok I can check about threads, but I cannot afford to enter into your code.
So for me is enough to know if listening to many ip then the socketlist of app() is one and unique.
Plz

**Reply by ghost** [reply]
Is the library single threaded? If so, does it make sense that there would be threads created? What happens if you print the thread id from within callbacks? Is it the same thread id as printed from main? These are interesting questions

**Reply by RDP1974** [reply]
thanks

**Comment by RDP1974** [comment]
last question then I will not bother anymore, I will build a prototype, then if successfull will pass through your commercial paying support

if I can see how to set SSL flags? ciphers, keys, considering that only AES openssl is cpu accelerated
export interface AppOptions {
    key_file_name?: RecognizedString;
    cert_file_name?: RecognizedString;
    ca_file_name?: RecognizedString;
    passphrase?: RecognizedString;
    dh_params_file_name?: RecognizedString;
    ssl_ciphers?: Rec

finally last question, ws remoteaddress I suppose will be "localhost" behind a proxy, how to ask X-HEADER HTTP before UPGRADE 101?

**Reply by uasan** [reply]
This returns an array of client IP addresses
```JS
Buffer.from(response.getRemoteAddressAsText()).toString().split(',')
```
